File: 92.cpp
// struct ListNode
// {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr)
//     {
//     }
//     ListNode(int x) : val(x), next(nullptr)
//     {
//     }
//     ListNode(int x, ListNode *next) : val(x), next(next)
//     {
//     }
// };
class Solution
{
  public:
    ListNode *reverseBetween(ListNode *head, int left, int right)
    {
        ListNode *dummy = new ListNode(0, head);
        ListNode *l1 = dummy;
        ListNode *l2 = dummy;
        for (int i = 0; i < left - 1; ++i)
        {
            l1 = l1->next;
        }
        for (int i = 0; i < right; ++i)
        {
            l2 = l2->next;
        }
        ListNode *curr = l1->next;
        ListNode *prev = nullptr;
        ListNode *next = nullptr;
        // disconnect ends
        ListNode *end = curr; 
        l1->next = nullptr;
        ListNode *right_side = l2->next;
        l2->next = nullptr;
        while (curr)
        {
            next = curr->next;
            curr->next = prev; 
            prev = curr; 
            curr = next; 
        }
        l1->next = prev; 
        end->next = right_side; 
        return dummy->next;
    }
};

File: 92.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right: 
            return head 
        dummy = ListNode(0, head)
        l, r = dummy, dummy
        for i in range(left - 1): 
            l = l.next 
        for i in range(right): 
            r = r.next
        # reverse
        curr = l.next 
        end = curr
        prev = None 
        nxt = None 
        # disconnect
        l.next = None
        right_side = r.next
        r.next = None
        # 
        while curr: 
            nxt = curr.next
            curr.next = prev 
            prev = curr 
            curr = nxt
        l.next = prev
        end.next = right_side
        return dummy.next
